File: ios-katha-ai-create-stories/KathaAICreateStories/Views/SubscriptionPaywall.swift
//
//  SubscriptionPaywall.swift
//  KathaAICreateStories
//

import SwiftUI

// MARK: - Subscription Paywall

struct SubscriptionPaywall: View {
    @Environment(AppState.self) private var appState
    @Environment(\.dismiss) private var dismiss
    @State private var selectedPlan: SubscriptionPlan = .yearly

    var body: some View {
        ZStack {
            // Background gradient
            LinearGradient(
                colors: [
                    KathaTheme.canvas,
                    KathaTheme.accentSoft.opacity(0.3),
                    KathaTheme.canvas
                ],
                startPoint: .top,
                endPoint: .bottom
            )
            .ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    closeBar
                    heroSection
                    featureList
                    planTiles
                    ctaSection
                    finePrint
                    SafeBottomSpacer()
                }
            }
            .scrollIndicators(.hidden)
        }
        .transition(.move(edge: .bottom).combined(with: .opacity))
    }

    // MARK: - Close Bar

    private var closeBar: some View {
        HStack {
            Button {
                Haptics.light()
                appState.closeSubscriptionPaywall()
            } label: {
                Image(systemName: "xmark")
                    .font(KathaFont.BodyStrong)
                    .foregroundStyle(KathaTheme.textPrimary)
                    .frame(width: 36, height: 36)
                    .background(
                        Circle().fill(KathaTheme.surface)
                    )
            }
            Spacer()
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.top, KathaTheme.Spacing.l)
    }

    // MARK: - Hero

    private var heroSection: some View {
        VStack(spacing: KathaTheme.Spacing.s) {
            Text("Katha")
                .font(KathaFont.Wordmark)
                .foregroundStyle(KathaTheme.accent)

            Text("PREMIUM")
                .font(KathaFont.PremiumWordmark)
                .foregroundStyle(KathaTheme.premium)
                .tracking(2)

            Text("Unlock unlimited storytelling.")
                .font(KathaFont.Title1)
                .foregroundStyle(KathaTheme.textPrimary)
                .multilineTextAlignment(.center)

            Text("For readers, writers, and everyone in between.")
                .font(KathaFont.Body)
                .foregroundStyle(KathaTheme.textSecondary)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, KathaTheme.Spacing.xxxl)
        .padding(.top, KathaTheme.Spacing.l)
    }

    // MARK: - Feature List

    private var featureList: some View {
        VStack(spacing: KathaTheme.Spacing.l) {
            ForEach(PremiumFeature.all) { feature in
                HStack(spacing: KathaTheme.Spacing.m) {
                    ZStack {
                        Circle()
                            .fill(KathaTheme.accent)
                            .frame(width: 24, height: 24)
                        Image(systemName: "checkmark")
                            .font(KathaFont.Meta)
                            .foregroundStyle(.white)
                    }

                    VStack(alignment: .leading, spacing: 2) {
                        Text(feature.title)
                            .font(KathaFont.BodyStrong)
                            .foregroundStyle(KathaTheme.textPrimary)
                        Text(feature.subtitle)
                            .font(KathaFont.Meta)
                            .foregroundStyle(KathaTheme.textSecondary)
                    }
                    Spacer()
                }
            }
        }
        .padding(.horizontal, KathaTheme.Spacing.xxxl)
        .padding(.top, KathaTheme.Spacing.xxxl)
    }

    // MARK: - Plan Tiles

    private var planTiles: some View {
        HStack(spacing: KathaTheme.Spacing.m) {
            // Monthly
            PlanTile(
                plan: .monthly,
                isSelected: selectedPlan == .monthly,
                isFeatured: false,
                onTap: {
                    Haptics.light()
                    withAnimation(.spring(duration: 0.2)) {
                        selectedPlan = .monthly
                    }
                }
            )
            .frame(width: UIScreen.main.bounds.width * 0.42)

            // Yearly
            PlanTile(
                plan: .yearly,
                isSelected: selectedPlan == .yearly,
                isFeatured: true,
                onTap: {
                    Haptics.light()
                    withAnimation(.spring(duration: 0.2)) {
                        selectedPlan = .yearly
                    }
                }
            )
            .frame(width: UIScreen.main.bounds.width * 0.46, height: 130)
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.top, KathaTheme.Spacing.xxxl)
    }

    // MARK: - CTA

    private var ctaSection: some View {
        VStack(spacing: KathaTheme.Spacing.s) {
            PremiumCTA(
                title: appState.isPurchasingSubscription
                    ? "Processing…"
                    : "Start Katha Premium · \(selectedPlan.price)\(selectedPlan == .yearly ? "/year" : "/month")"
            ) {
                Task { await appState.purchaseSubscription(selectedPlan) }
            }
            .disabled(appState.isPurchasingSubscription)
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.top, KathaTheme.Spacing.xxxl)
    }

    // MARK: - Fine Print

    private var finePrint: some View {
        VStack(spacing: KathaTheme.Spacing.m) {
            Text("Cancel anytime in Settings. Renews automatically unless cancelled 24 hours before period end. Payment charged to your Apple ID or Google account.")
                .font(KathaFont.Meta)
                .foregroundStyle(KathaTheme.textTertiary)
                .multilineTextAlignment(.center)

            if appState.isRestoringPurchases {
                HStack(spacing: 6) {
                    ProgressView()
                        .tint(KathaTheme.accent)
                        .scaleEffect(0.7)
                    Text("Restoring…")
                        .font(KathaFont.Caption)
                        .foregroundStyle(KathaTheme.accent)
                }
            } else {
                TextLink(title: "Restore purchases") {
                    Task { await appState.restorePurchases() }
                }
            }

            HStack(spacing: 0) {
                TextLink(title: "Terms") { appState.showToast("Terms: katha.ai/terms") }
                Text(" · ")
                    .font(KathaFont.Caption)
                    .foregroundStyle(KathaTheme.textTertiary)
                TextLink(title: "Privacy") { appState.showToast("Privacy: katha.ai/privacy") }
            }
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.top, KathaTheme.Spacing.m)
    }
}

// MARK: - Plan Tile

struct PlanTile: View {
    let plan: SubscriptionPlan
    let isSelected: Bool
    let isFeatured: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            VStack(spacing: KathaTheme.Spacing.xs) {
                Text(plan.displayName)
                    .font(KathaFont.Meta)
                    .foregroundStyle(isFeatured && isSelected ? KathaTheme.premium : KathaTheme.textSecondary)
                    .tracking(1)

                Text(plan.price)
                    .font(KathaFont.Title1)
                    .foregroundStyle(KathaTheme.textPrimary)

                Text(plan.perMonth)
                    .font(KathaFont.Meta)
                    .foregroundStyle(KathaTheme.textTertiary)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(
                RoundedRectangle(cornerRadius: KathaTheme.Radius.l)
                    .fill(isSelected ? (isFeatured ? KathaTheme.premiumSoft.opacity(0.3) : KathaTheme.accentSoft.opacity(0.3)) : KathaTheme.surface)
                    .overlay(
                        RoundedRectangle(cornerRadius: KathaTheme.Radius.l)
                            .stroke(isSelected ? (isFeatured ? KathaTheme.premium : KathaTheme.accent) : KathaTheme.border, lineWidth: isSelected ? 2 : 1.5)
                    )
            )
            .overlay(alignment: .topTrailing) {
                if isFeatured {
                    Text("SAVE 40%")
                        .font(KathaFont.Meta)
                        .foregroundStyle(.white)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 3)
                        .background(Capsule().fill(KathaTheme.premium))
                        .padding(8)
                }
            }
            .overlay(alignment: .topLeading) {
                if isSelected {
                    ZStack {
                        Circle()
                            .fill(isFeatured ? KathaTheme.premium : KathaTheme.accent)
                            .frame(width: 20, height: 20)
                        Image(systemName: "checkmark")
                            .font(KathaFont.Meta)
                            .foregroundStyle(.white)
                    }
                    .padding(8)
                }
            }
        }
        .buttonStyle(PressScaleStyle(scale: 0.98))
    }
}

// MARK: - Premium CTA (inverted style for paywall)

struct PremiumCTAStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(configuration.isPressed ? KathaTheme.premium.opacity(0.8) : KathaTheme.premium)
            )
            .shadow(
                color: KathaTheme.premium.opacity(configuration.isPressed ? 0.15 : 0.3),
                radius: configuration.isPressed ? 5 : 8,
                y: configuration.isPressed ? 1 : 2
            )
            .animation(.easeOut(duration: 0.1), value: configuration.isPressed)
    }
}

File: android-katha-ai/app/src/main/java/com/rork/kathaai/ui/screens/SubscriptionPaywall.kt
package com.rork.kathaai.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.WorkspacePremium
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.kathaai.model.PremiumFeature
import com.rork.kathaai.model.SubscriptionPlan
import com.rork.kathaai.ui.components.PrimaryCTA
import com.rork.kathaai.ui.components.SafeBottomSpacer
import com.rork.kathaai.ui.components.TextLink
import com.rork.kathaai.ui.theme.KathaTheme
import com.rork.kathaai.ui.theme.serif
import com.rork.kathaai.viewmodel.AppViewModel
import com.rork.kathaai.viewmodel.KathaUiState

@Composable
fun SubscriptionPaywall(
    state: KathaUiState,
    viewModel: AppViewModel,
    onClose: () -> Unit
) {
    var selectedPlan by remember { mutableStateOf(SubscriptionPlan.YEARLY) }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                Brush.verticalGradient(
                    listOf(KathaTheme.canvas, KathaTheme.accentSoft.copy(alpha = 0.3f), KathaTheme.canvas)
                )
            )
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(bottom = 120.dp)
        ) {
            // Close bar
            Row(
                modifier = Modifier.fillMaxWidth().padding(start = 20.dp, top = 20.dp),
                horizontalArrangement = Arrangement.Start
            ) {
                Box(
                    modifier = Modifier
                        .size(36.dp)
                        .clip(CircleShape)
                        .background(KathaTheme.surface)
                        .clickable { onClose() },
                    contentAlignment = Alignment.Center
                ) {
                    Icon(Icons.Outlined.Close, "Close", tint = KathaTheme.textPrimary, modifier = Modifier.size(16.dp))
                }
            }

            // Hero
            Column(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 32.dp, vertical = 20.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text("Katha", style = serif(24, FontWeight.Bold), color = KathaTheme.accent)
                Text("PREMIUM", style = serif(32, FontWeight.Bold), color = KathaTheme.premium)
                Text("Unlock unlimited storytelling.", fontSize = 28.sp, fontWeight = FontWeight.Bold, color = KathaTheme.textPrimary, textAlign = TextAlign.Center)
                Text("For readers, writers, and everyone in between.", fontSize = 15.sp, color = KathaTheme.textSecondary, textAlign = TextAlign.Center)
            }

            // Features
            Column(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 32.dp, vertical = 16.dp),
                verticalArrangement = Arrangement.spacedBy(20.dp)
            ) {
                PremiumFeature.all.forEach { feature ->
                    Row(horizontalArrangement = Arrangement.spacedBy(16.dp), verticalAlignment = Alignment.CenterVertically) {
                        Box(
                            modifier = Modifier.size(24.dp).clip(CircleShape).background(KathaTheme.accent),
                            contentAlignment = Alignment.Center
                        ) {
                            Icon(Icons.Outlined.Check, null, tint = Color.White, modifier = Modifier.size(12.dp))
                        }
                        Column {
                            Text(feature.title, fontSize = 15.sp, fontWeight = FontWeight.SemiBold, color = KathaTheme.textPrimary)
                            Text(feature.subtitle, fontSize = 12.sp, color = KathaTheme.textSecondary)
                        }
                    }
                }
            }

            // Plan tiles
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 32.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                PlanTile(
                    plan = SubscriptionPlan.MONTHLY,
                    isSelected = selectedPlan == SubscriptionPlan.MONTHLY,
                    isFeatured = false,
                    modifier = Modifier.weight(0.42f),
                    onClick = { selectedPlan = SubscriptionPlan.MONTHLY }
                )
                PlanTile(
                    plan = SubscriptionPlan.YEARLY,
                    isSelected = selectedPlan == SubscriptionPlan.YEARLY,
                    isFeatured = true,
                    modifier = Modifier.weight(0.46f),
                    onClick = { selectedPlan = SubscriptionPlan.YEARLY }
                )
            }

            // CTA
            val ctaLabel = if (state.isPurchasingSubscription) "Processing…" else "Start Katha Premium · ${selectedPlan.price}${if (selectedPlan == SubscriptionPlan.YEARLY) "/year" else "/month"}"
            PrimaryCTA(title = ctaLabel) {
                viewModel.purchaseSubscription(selectedPlan)
            }

            // Fine print
            Column(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 12.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text("Cancel anytime in Settings. Renews automatically unless cancelled 24 hours before period end. Payment charged to your Google account.", fontSize = 11.sp, color = KathaTheme.textTertiary, textAlign = TextAlign.Center)
                if (state.isRestoringPurchases) {
                    Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                        CircularProgressIndicator(modifier = Modifier.size(12.dp), color = KathaTheme.accent, strokeWidth = 1.5.dp)
                        Text("Restoring…", fontSize = 13.sp, color = KathaTheme.accent)
                    }
                } else {
                    TextLink(title = "Restore purchases") { viewModel.restorePurchases() }
                }
                Row(horizontalArrangement = Arrangement.spacedBy(0.dp)) {
                    TextLink(title = "Terms") { viewModel.showToast("Terms: katha.ai/terms") }
                    Text(" · ", fontSize = 13.sp, color = KathaTheme.textTertiary)
                    TextLink(title = "Privacy") { viewModel.showToast("Privacy: katha.ai/privacy") }
                }
            }
        }
    }
}

@Composable
private fun PlanTile(
    plan: SubscriptionPlan,
    isSelected: Boolean,
    isFeatured: Boolean,
    modifier: Modifier = Modifier,
    onClick: () -> Unit
) {
    val borderColor = if (isSelected) { if (isFeatured) KathaTheme.premium else KathaTheme.accent } else KathaTheme.border
    val bg = if (isSelected) { if (isFeatured) KathaTheme.premiumSoft.copy(alpha = 0.3f) else KathaTheme.accentSoft.copy(alpha = 0.3f) } else KathaTheme.surface

    Box(
        modifier = modifier
            .height(if (isFeatured) 130.dp else 100.dp)
            .clip(RoundedCornerShape(KathaTheme.Radius.l))
            .background(bg)
            .clickable { onClick() }
    ) {
        Column(
            modifier = Modifier.fillMaxSize().padding(vertical = 8.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Text(plan.displayName, fontSize = 12.sp, fontWeight = FontWeight.Bold, color = if (isFeatured && isSelected) KathaTheme.premium else KathaTheme.textSecondary)
            Text(plan.price, fontSize = 28.sp, fontWeight = FontWeight.Bold, color = KathaTheme.textPrimary)
            Text(plan.perMonth, fontSize = 11.sp, color = KathaTheme.textTertiary)
        }
        if (isFeatured) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(8.dp)
                    .clip(RoundedCornerShape(50))
                    .background(KathaTheme.premium)
                    .padding(horizontal = 6.dp, vertical = 3.dp)
            ) {
                Text("SAVE 40%", fontSize = 9.sp, fontWeight = FontWeight.Bold, color = Color.White)
            }
        }
        if (isSelected) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopStart)
                    .padding(8.dp)
                    .size(20.dp)
                    .clip(CircleShape)
                    .background(if (isFeatured) KathaTheme.premium else KathaTheme.accent),
                contentAlignment = Alignment.Center
            ) {
                Icon(Icons.Outlined.Check, null, tint = Color.White, modifier = Modifier.size(10.dp))
            }
        }
    }
}
